### PR TITLE
New resource `azurerm_disk_pool_iscsi_target` base on `azurerm_disk_pool` on pr #14675

### PR DESCRIPTION
**Pull request by lonegunmanb**
Base on `azurerm_disk_pool`, added new resource `azurerm_disk_pool_iscsi_target` for disk pool.

This resource HAVE NOT embedded lun arguments, I think it'll be better to make a new seperate resource to managed lun disk in an iscsi target, otherwise the user need delete embedded lun inside iscsi target resource before they can delete the managed disk and the attachment.

Acc tests:

=== RUN   TestDiskPoolIscsiTarget_basic
=== PAUSE TestDiskPoolIscsiTarget_basic
=== CONT  TestDiskPoolIscsiTarget_basic
--- PASS: TestDiskPoolIscsiTarget_basic (1652.70s)
=== RUN   TestDiskPoolIscsiTarget_requiresImport
=== PAUSE TestDiskPoolIscsiTarget_requiresImport
=== CONT  TestDiskPoolIscsiTarget_requiresImport
--- PASS: TestDiskPoolIscsiTarget_requiresImport (1660.95s)
PASS

Process finished with the exit code 0

### CODE REVIEW

**Comment by lonegunmanb** [comment]
Added retry logic to iscsi target resource, improved disk pool resource's retryable error's readability.
Acc tests:

=== RUN   TestAccDiskPool_basic
=== PAUSE TestAccDiskPool_basic
=== CONT  TestAccDiskPool_basic
--- PASS: TestAccDiskPool_basic (984.25s)
=== RUN   TestAccDiskPool_requiresImport
=== PAUSE TestAccDiskPool_requiresImport
=== CONT  TestAccDiskPool_requiresImport
--- PASS: TestAccDiskPool_requiresImport (1066.95s)
=== RUN   TestAccDiskPool_complete
=== PAUSE TestAccDiskPool_complete
=== CONT  TestAccDiskPool_complete
--- PASS: TestAccDiskPool_complete (1474.21s)
PASS



=== RUN   TestAccDiskPoolDiskAttachment_basic
=== PAUSE TestAccDiskPoolDiskAttachment_basic
=== CONT  TestAccDiskPoolDiskAttachment_basic
--- PASS: TestAccDiskPoolDiskAttachment_basic (1701.48s)
=== RUN   TestAccDiskPoolDiskAttachment_requiresImport
=== PAUSE TestAccDiskPoolDiskAttachment_requiresImport
=== CONT  TestAccDiskPoolDiskAttachment_requiresImport
--- PASS: TestAccDiskPoolDiskAttachment_requiresImport (1445.02s)
=== RUN   TestAccDiskPoolDiskAttachment_multipleDisks
=== PAUSE TestAccDiskPoolDiskAttachment_multipleDisks
=== CONT  TestAccDiskPoolDiskAttachment_multipleDisks
--- PASS: TestAccDiskPoolDiskAttachment_multipleDisks (2322.07s)
=== RUN   TestAccDiskPoolDiskAttachment_destroy
=== PAUSE TestAccDiskPoolDiskAttachment_destroy
=== CONT  TestAccDiskPoolDiskAttachment_destroy
--- PASS: TestAccDiskPoolDiskAttachment_destroy (1693.86s)
PASS

=== RUN   TestDiskPoolIscsiTarget_basic
=== PAUSE TestDiskPoolIscsiTarget_basic
=== CONT  TestDiskPoolIscsiTarget_basic
--- PASS: TestDiskPoolIscsiTarget_basic (1799.89s)
=== RUN   TestDiskPoolIscsiTarget_requiresImport
=== PAUSE TestDiskPoolIscsiTarget_requiresImport
=== CONT  TestDiskPoolIscsiTarget_requiresImport
--- PASS: TestDiskPoolIscsiTarget_requiresImport (1694.25s)
PASS

File: internal/services/disks/disk_pool_iscsi_target_resource.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+			if err != nil {
+				return err
+			}
+			if poolId.SubscriptionId != metadata.Client.Account.SubscriptionId {
+				return fmt.Errorf("Disk Pool subscription id %q is different from provider's subscription", poolId.SubscriptionId)
+			}
+			id := iscsitargets.NewIscsiTargetID(poolId.SubscriptionId, poolId.ResourceGroupName, poolId.DiskPoolName, m.Name)
+			client := metadata.Client.Disks.DisksPoolIscsiTargetClient
+			locks.ByID(poolId.ID())
+			defer locks.UnlockByID(poolId.ID())
```
Comment (katbyte) [inline internal/services/disks/disk_pool_iscsi_target_resource.go:143]: could we add whitespace throughout this PR for readability 
```suggestion
			m := DiskPoolIscsiTargetModel{}
			err := metadata.Decode(&m)
			if err != nil {
				return err
			}
			
			poolId, err := iscsitargets.ParseDiskPoolID(m.DisksPoolId)
			if err != nil {
				return err
			}
			if poolId.SubscriptionId != metadata.Client.Account.SubscriptionId {
				return fmt.Errorf("Disk Pool subscription id %q is different from provider's subscription", poolId.SubscriptionId)
			}
			
			id := iscsitargets.NewIscsiTargetID(poolId.SubscriptionId, poolId.ResourceGroupName, poolId.DiskPoolName, m.Name)
			client := metadata.Client.Disks.DisksPoolIscsiTargetClient
			locks.ByID(poolId.ID())
			defer locks.UnlockByID(poolId.ID())
```

etc

File: internal/services/disks/disk_pool_iscsi_target_resource.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+				false,
+			),
+		},
+		"disks_pool_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: iscsitargets.ValidateDiskPoolID,
+		},
+		"lun": {
```
Comment (katbyte) [inline internal/services/disks/disk_pool_iscsi_target_resource.go:64]: and here
```suggestion
		},
		
		"acl_mode": {
			Type:     pluginsdk.TypeString,
			Required: true,
			ForceNew: true,
			ValidateFunc: validation.StringInSlice(
				[]string{string(iscsitargets.IscsiTargetAclModeDynamic)},
				false,
			),
		},
		
		"disks_pool_id": {
			Type:         pluginsdk.TypeString,
			Required:     true,
			ForceNew:     true,
			ValidateFunc: iscsitargets.ValidateDiskPoolID,
		},
		
		"lun": {
```

File: website/docs/r/disk_pool_iscsi_target.html.markdown [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+The following arguments are supported:
+
+* `acl_mode` - (Required) Mode for Target connectivity. The only supported value is `Dynamic` for now. Changing this forces a new iSCSI Target to be created.
+
+* `disks_pool_id` - (Required) The ID of the Disk Pool. Changing this forces a new iSCSI Target to be created.
+
+* `name` - (Required) The name of the iSCSI Target. The name can only contain lowercase letters, numbers, periods, or hyphens, and length should between [5-223]. Changing this forces a new iSCSI Target to be created.
+
+* `target_iqn` - (Optional) ISCSI Target IQN (iSCSI Qualified Name); example: "iqn.2005-03.org.iscsi:server". IQN should follow the format `iqn.yyyy-mm.<abc>.<pqr>[:xyz]`; supported characters include [0-9a-z-.:], and the length should between [4-223]. Changing this forces a new iSCSI Target to be created.
```
Comment (katbyte) [inline website/docs/r/disk_pool_iscsi_target.html.markdown:105]: could we properly quote values
```suggestion
* `target_iqn` - (Optional) ISCSI Target IQN (iSCSI Qualified Name); example: `iqn.2005-03.org.iscsi:server`. IQN should follow the format `iqn.yyyy-mm.<abc>.<pqr>[:xyz]`; supported characters include [0-9a-z-.:], and the length should between [4-223]. Changing this forces a new iSCSI Target to be created.
```

and not use regex here? property sentences please ie 4-223 should be "between `4` and `223`, same for  [0-9a-z-.:]

**Review by katbyte — CHANGES_REQUESTED** [review]

**Review by katbyte — APPROVED** [review]
Thanks @lonegunmanb - LGTM 💯

**Pull request merged**